File: examples/exo/cpp/src/imu.cpp
#include "exo_example_support.hpp"

#include <cstdlib>
#include <exception>
#include <iostream>
#include <optional>
#include <string>

namespace {

void print_imu(const std::string& label,
               const std::optional<qnbot::ExoImuState>& imu) {
    if (!imu) return;
    std::cout << label << " acceleration=";
    example::print_values(imu->acceleration_mps2);
    std::cout << " angular_velocity=";
    example::print_values(imu->angular_velocity_rad_s);
    std::cout << " orientation=";
    example::print_values(imu->orientation_xyzw);
    std::cout << '\n';
}

} // namespace

int main(int argc, char** argv) {
    try {
        std::string port;
        std::string name = "primary";
        for (int index = 1; index < argc; ++index) {
            const std::string argument = argv[index];
            if (argument == "--port") {
                port = example::require_value(argc, argv, index, argument);
            } else if (argument == "--name") {
                name = example::require_value(argc, argv, index, argument);
            } else {
                throw std::invalid_argument("unknown argument: " + argument);
            }
        }
        if (port.empty()) throw std::invalid_argument("--port is required");

        qnbot::Sdk sdk(example::serial_config(port, name));
        auto exo = sdk.exo();
        auto device = exo.device(name);
        if (!device.get_device_info().capabilities.imus) {
            std::cout << "this device reports no IMU" << std::endl;
            exo.close();
            return EXIT_SUCCESS;
        }

        auto telemetry = device.telemetry();
        const auto subscription = telemetry.subscribe(
            [](const qnbot::Sample<qnbot::ExoTelemetry>& sample) {
                print_imu("torso", sample.value.payload.torso_imu);
                print_imu("extra", sample.value.payload.extra_imu);
            });

        exo.start();
        std::cout << "running; press Ctrl+C to stop" << std::endl;
        exo.run_forever();
        static_cast<void>(subscription);
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}

File: exo/python/src/imu.py
from __future__ import annotations

import argparse

from qnbot_sdk import Sample, Sdk, SerialConnection
from qnbot_sdk.exo import ExoConfig, ExoTelemetry


def create_sdk(port: str, name: str) -> Sdk:
    return Sdk(devices=(ExoConfig(name=name, connection=SerialConnection(port=port)),))


def print_imus(sample: Sample[ExoTelemetry]) -> None:
    payload = sample.value.payload
    for label, imu in (("torso", payload.torso_imu), ("extra", payload.extra_imu)):
        if imu is None:
            continue
        print(
            f"{label} imu sequence={sample.sequence} "
            f"acceleration={imu.acceleration_mps2} "
            f"angular_velocity={imu.angular_velocity_rad_s} "
            f"orientation={imu.orientation_xyzw}"
        )


def main() -> None:
    arguments = argparse.ArgumentParser(description="Read Exo IMU payloads")
    arguments.add_argument("--port", required=True, help="Serial port for the device")
    arguments.add_argument("--name", default="primary", help="Logical device name")
    options = arguments.parse_args()

    sdk = create_sdk(options.port, options.name)
    exo = sdk.exo()
    device = exo.device(options.name)
    if not device.get_device_info().capabilities.imus:
        print("this device reports no IMU", flush=True)
        exo.close()
        return

    device.telemetry().subscribe(print_imus)
    exo.start()
    print("running; press Ctrl+C to stop", flush=True)
    exo.run_forever()


if __name__ == "__main__":
    main()
